cephadm: add locking

liewegas opened the pull request.

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -1188,6 +1381,9 @@ def command_bootstrap():
     mgr_id = args.mgr_id or generate_service_id()
     logging.info('Cluster fsid: %s' % fsid)
 
+    l = FileLock(fsid + '.deploy')
```
add `.lock` or similar to the filename for clarity?

mgfritch commented on src/cephadm/cephadm:
```diff
@@ -1499,6 +1695,9 @@ def command_deploy():
     # type: () -> None
     (daemon_type, daemon_id) = args.name.split('.', 1)
 
+    l = FileLock(args.fsid + '.deploy')
```
add `<service>.<id>` to the filename? .. to allow deploying OSDs etc in parallel ..

sebastian-philipp replied:
well - actually we shouldn't allow deploying osds in parallel at all. No idea what happens, if we deploy osds using two different drive groups on one system concurrently. 

On the other hand, we should allow running `device ls` and `service ls` and deploying (osd|mon|mgr|...)s in parallel

liewegas replied:
I think we should stuck to a global deploy lock for now.  otherwise, we need to introduce read/write shared locks, since you want to allow >1 mds/rgw/wahtever deploys to proceed, but those should still conflict with (say) an rm-cluster.  I don't think it's worth the complexity.  The most important goal right now is to get parallelism across multiple hosts; less important on a single host--we can always improve that later.

sebastian-philipp commented on src/cephadm/cephadm:
```diff
@@ -1894,6 +2097,10 @@ def command_adopt():
     # type: () -> None
+
+    l = FileLock(args.fsid + '.deploy')
+    l.acquire()
```
when in the file lock released?

liewegas replied:
when the object is destroyed

liewegas commented:
repushed.
- fixed fsid for adopt
- added .lock to filename
- shorted all lots to just the fsid.  i think this will be nicer later if/when we add read/write and granular locks.. .e.g. deploying mds.a might lock $fsid (shared) and $fsid.mds.a (exclusive).  But for now, all locks are exclusive.
